### PR TITLE
TrajoptPlanner throws (Self)CollisionPlanningError and ConstraintViol…

### PR DESCRIPTION
**Pull request by gilwoolee**
…ationPlanningError whenever possible.

### CODE REVIEW

File: src/or_trajopt/planning.py [hunk 1]
```diff
@@ -632,3 +634,45 @@ def _WaypointsToTraj(self, robot, waypoints):
...
+
+        p = openravepy.KinBody.SaveParameters
+        with robot.CreateRobotStateSaver(p.ActiveDOF):
+            self._checkCollisionForIKSolutions(robot, ik_solutions)
+
+        raise PlanningError('No collision-free IK solution.')
+
+    def _checkCollisionForIKSolutions(self, robot, ik_solutions): 
+        """ Raise collision error if there is one in ik_solutions
+        Should be called while saving robot's curent state
```
Comment (psigen) [inline src/or_trajopt/planning.py]: typo: "current"

File: src/or_trajopt/planning.py [hunk 2]
```diff
@@ -14,6 +14,9 @@
                                 PlanningError,
                                 PlanningMethod,
                                 Tags)
+from prpy.planning.exceptions import (CollisionPlanningError,
+                                    SelfCollisionPlanningError,
```
Comment (psigen) [inline src/or_trajopt/planning.py]: Fix indentation to match paren in line above.

File: src/or_trajopt/planning.py [hunk 3]
```diff
@@ -560,11 +565,8 @@ def PlanToTSR(self, robot, tsrchains, is_interactive=False, **kw_args):
         with robot.CreateRobotStateSaver(p.ActiveDOF):
             # Set robot DOFs to DOFs in optimization problem
             prob.SetRobotActiveDOFs()
-
             # Check that trajectory is collision free
-            from trajoptpy.check_traj import traj_is_safe
-            if not traj_is_safe(waypoints, robot):
-                return PlanningError("Result was in collision.")
+            self._checkCollisionForIKSolutions(robot, waypoints)
```
Comment (psigen) [inline src/or_trajopt/planning.py]: I'm not sure this is the correct logic here.  This only checks for collision immediately at each waypoint, but the trajectory could collide in between waypoints.

@mkoval: Does `prpy` have a utility function to check if an entire trajectory is collision free at the active DOF resolution?

**Comment by psigen** [comment]
Thanks Gilwoo!  I added a few comments, but overall the new exception handling looks much more helpful.

File: src/or_trajopt/planning.py (hunk 3 above)
Comment (mkoval) [inline src/or_trajopt/planning.py]: You can call `util.GetCollisionCheckPts` to get a list of joint configurations. Then, for each configuration, you can perform whatever tests you want, e.g. collision, self-collision, constraint violation, etc.

**Comment by gilwoolee** [comment]
@psigen @mkoval  Fixed collision check routine to use util.GetCollisionCheckPts. Fixed indentation & typo as well.

File: src/or_trajopt/planning.py [hunk 4]
```diff
@@ -281,9 +285,13 @@ def _Plan(self, robot, request,
...
+                traj = self._WaypointsToTraj(robot, waypoints)
+                # Generate timed trajectory, required for GetCollisionCheckPTs
+                traj = util.ComputeUnitTiming(robot, traj)
+                checkpoints = util.GetCollisionCheckPts(robot, traj,
+                        include_start=True)
+                for t_check, q_check in checkpoints:
+                    self._checkCollisionForIKSolutions(robot, [q_check])
 
             # Convert the waypoints to a trajectory.
             path = self._WaypointsToTraj(robot, waypoints)
```
Comment (mkoval) [inline src/or_trajopt/planning.py]: We shouldn't need to compute `_WaypointsToTraj` twice (lines 288 and 297).

File: src/or_trajopt/planning.py [hunk 5]
```diff
@@ -560,11 +572,14 @@ def PlanToTSR(self, robot, tsrchains, is_interactive=False, **kw_args):
...
+            traj = self._WaypointsToTraj(robot, waypoints)
+            # Generate timed trajectory, required for GetCollisionCheckPTs
+            traj = util.ComputeUnitTiming(robot, traj) 
+            checkpoints = util.GetCollisionCheckPts(robot, traj,
+                        include_start=True)
+            for t_check, q_check in checkpoints:
+                self._checkCollisionForIKSolutions(robot, [q_check])
 
         # Convert the waypoints to a trajectory.
         return self._WaypointsToTraj(robot, waypoints)
```
Comment (mkoval) [inline src/or_trajopt/planning.py]: Again, it would be ideal to not call `_WaypointsToTraj` twice.

File: src/or_trajopt/planning.py [hunk 6]
```diff
@@ -632,3 +647,45 @@ def _WaypointsToTraj(self, robot, waypoints):
...
+
+        ik_solutions = manipulator.FindIKSolutions(
+            ik_param,
+            openravepy.IkFilterOptions.IgnoreSelfCollisions,
+            ikreturn=False,
+            releasegil=True
+            )
+
+        p = openravepy.KinBody.SaveParameters
+        with robot.CreateRobotStateSaver(p.ActiveDOF):
```
Comment (mkoval) [inline src/or_trajopt/planning.py]: I think this is intended to be `p.LinkTransformations`, which saves the active DOF **values**. `ActiveDOF` saves the active DOF **indices**, which you are not changing in this code.

Yes, the names are very confusing. :sweat:

Comment (mkoval) [inline src/or_trajopt/planning.py]: Also, I would prefer that you move this inside `_checkCollisionForIKSolutions`, where the `SetActiveDOFValues` call actually happens.

File: src/or_trajopt/planning.py [hunk 7]
```diff
@@ -632,3 +647,45 @@ def _WaypointsToTraj(self, robot, waypoints):
...
+            self._checkCollisionForIKSolutions(robot, ik_solutions)
+
+        raise PlanningError('No collision-free IK solution.')
```
Comment (mkoval) [inline src/or_trajopt/planning.py]: Unless there is a bug we should never hit this point, correct?

If so, you should change to a generic Python exception class (like `Exception`) so it is not caught (and potentially hidden) by a `MetaPlanner`.

File: src/or_trajopt/planning.py [hunk 8]
```diff
@@ -547,8 +560,7 @@ def PlanToTSR(self, robot, tsrchains, is_interactive=False, **kw_args):
 
         # Check for constraint violations.
         if result.GetConstraints():
-            raise PlanningError("Trajectory did not satisfy constraints: {:s}"
-                                .format(str(result.GetConstraints())))
+            raise ConstraintViolationPlanningError(str(result.GetConstraints()[0]))
```
Comment (mkoval) [inline src/or_trajopt/planning.py]: Is it possible to pass `violation_by` here, as well?

Comment (gilwoolee) [inline src/or_trajopt/planning.py]: Hmm. I can, but I don't have a good test case for this yet.

**Comment by mkoval** [comment]
@psigen Have you reviewed this yet?

**Comment by mkoval** [comment]
@psigen Could you please review this soon? I'd like to merge https://github.com/personalrobotics/prpy/pull/221, but it's blocked by this pull request.

File: src/or_trajopt/planning.py [hunk 9]
```diff
@@ -547,8 +560,10 @@ def PlanToTSR(self, robot, tsrchains, is_interactive=False, **kw_args):
 
         # Check for constraint violations.
         if result.GetConstraints():
-            raise PlanningError("Trajectory did not satisfy constraints: {:s}"
-                                .format(str(result.GetConstraints())))
+            # Pick the first violate constraint
+            name = result.GetConstraints()[0][0]
+            error = result.GetConstraints()[0][1]
+            raise ConstraintViolationPlanningError(name, vioaltion_by = error)
```
Comment (psigen) [inline src/or_trajopt/planning.py]: typo in argument `violation_by`?

**Pull request merged**